third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/MemoryOpToLLVM.cpp

```diff
@@ -136,64 +144,149 @@ struct LocalLoadOpConversion
-  if (!mmaEncoding)
+  if (targetInfo.getComputeCapability() < 90)
     return failure();
-  auto sharedLayout =
-      dyn_cast<triton::gpu::NVMMASharedEncodingAttr>(memDescType.getEncoding());
-  if (!sharedLayout)
+
+  assert(llvmOpCount == nullptr && "NYI");
+  auto b = TritonLLVMOpBuilder(loc, rewriter);
+  auto *ctx = tensorTy.getContext();
+  auto regL = toLinearLayout(tensorTy.getShape(), tensorTy.getEncoding());
+  auto memL = toLinearLayout(memDescType.getShape(), memDescType.getEncoding());
+  auto cvt = minimalCvtLayout(memDescType, tensorTy);
+
+  auto S = [ctx](StringRef v) { return StringAttr::get(ctx, v); };
+  auto kReg = S("register");
+  auto kLane = S("lane");
+  auto kWarp = S("warp");
+  auto kBlock = S("block");
+  auto kOffset = S("offset");
+  auto smemPtrTy = ptr_ty(ctx, 3);
+
+  // Just stmatrix for now
+  // 1) NYI in the stmatrix lowering
+  auto bitwidth = tensorTy.getElementTypeBitWidth();
+  if (bitwidth != 16)
     return failure();
-  int swizzleByteSize = sharedLayout.getSwizzlingByteWidth();
 
-  RankedTensorType srcTy = src.getType();
+  // NYI.
+  if (regL.getInDimSize(kBlock) != 1)
+    return failure();
 
-  SmallVector<unsigned> shape =
-      convertType<unsigned, int64_t>(srcTy.getShape());
-  SmallVector<unsigned> order = sharedLayout.getTransposed()
-                                    ? SmallVector<unsigned>({0, 1})
-                                    : SmallVector<unsigned>({1, 0});
-  if (!targetInfo.canUseStMatrix(srcTy, shape, shape, order, swizzleByteSize)) {
+  // Compute quotient
+  auto tile = LinearLayout::identity1D(32 / bitwidth, kReg, kOffset) *
+              LinearLayout::identity1D(4, kLane, kOffset);
+  auto maybeQuot = divideLeft(cvt, tile);
+  if (!maybeQuot.has_value()) {
     return failure();
   }
+  auto quot = maybeQuot.value();
 
-  auto *ctx = rewriter.getContext();
+  // Compute the registers that don't broadcast to see how many elements we can
+  // vectorise over
+  auto mask =
+      (~quot.getFreeVariableMasks()[kReg]) & (quot.getInDimSize(kReg) - 1);
+  auto vec = std::min(2, __builtin_popcount(mask));
+  // 2) NYI. stmatrix.x4 is the only supported for now
+  if (vec != 2) {
+    return failure();
+  }
 
-  auto layout =
-      chooseStMatrixLayout(rewriter.getContext(), srcTy, swizzleByteSize);
-  auto llvmElemTy = typeConverter->convertType(memDescType.getElementType());
-  auto smemPtrTy = ptr_ty(ctx, 3);
+  uint32_t newMask = 0;
+  for (int i = 0; i < vec; ++i) {
+    auto lowest = mask & -mask;
+    newMask |= lowest;
+    mask -= lowest;
+  }
+  mask = newMask;
 
-  auto kRegister = str_attr("register");
-  auto kLane = str_attr("lane");
-  auto kWarp = str_attr("warp");
-  auto kBlock = str_attr("block");
+  auto srcVals = unpackLLElements(loc, adaptorSrc, rewriter);
+  // If the permutation is non-trivial, we need to permute the layout and values
+  if (!llvm::isPowerOf2_32(mask + 1)) {
```

Comment:
IIUC this is handling cases where we may be have duplicated values by permuting to bring contiguous values together, e.g. [a, a, b,  b, c, c, d, d] -> [a, b, c, d, a, b, c, d] now we can use a x4 store for contiguous register indexes.

Would it also work to have a helper function:
```
auto [reducedSrcValues, reducedLayout] = removeDuplicateValues(srcValues, Layout);
```
Which returns you `[a, b, c, d]` and the corresponding layout with fewer register dims. With this, I think the code would be a lot more readable.

Reply:
Looks much better now, thanks!